### DISCUSSION TITLE
harmonize TaskFlow and custom operator usage

### DISCUSSION THREAD

**Discussion by stephenonethree** [apache/airflow#27038] [Ideas]
First, let me describe my motivation. I've been watching the new TaskFlow API from afar and reading about it, and it seems super cool. I like how it simplifies things. However I haven't been able to use it because all of my operators are custom. My code is fairly complex and it helps to have an object-oriented approach. As a result I'm also a bit nervous, because it seems like Airflow now has two very different ways of doing things, and I'm worried about getting left behind.

So, I was thinking about if/how these two approaches could be harmonized. I think it makes sense to keep the **declarations** separate. TaskFlow deals with code that can be expressed as a function, while custom operators deal with code that is expressed as a class. Different tools for different use cases.

However, it seems like the usage of the objects once they are instantiated should be the same. As a user, I shouldn't need to care if a Task was created with TaskFlow or a custom operator. It should have the same functionality.

Right now you get separate types depending on how they were defined. A TaskFlow object is an `XComArg`, while a custom object is whatever the custom class is. However, I would argue that in practice they are all just Tasks. Even an `XComArg` supports the ` >> ` operation for defining dependencies, which shows that in practice it isn't just for representing XCom values, it's basically still a Task.

You can see the downsides of having different types right now, if you look at the [Dynamic Task Mapping documentation](https://airflow.apache.org/docs/apache-airflow/stable/concepts/dynamic-task-mapping.html). There are 4 different ways to combine classic and TaskFlow objects (classic-classic, TaskFlow-TaskFlow, classic-TaskFlow, TaskFlow-classic). If the types were harmonized this could be greatly simplified because there'd only be one way.

I'm not sure what the best/easiest way of doing this is. Maybe it would be possible to annotate classic operator classes so that they generate an XComArg as well. I'm not so familiar with annotations though.

Basically, I'm recalling the Python motto: "There's only one way to do it."

I'm sorry to propose this because it might be a lot of work, but maybe there's an easy way. I do think it will make things easier in the long run, for users and probably maintainers.

**Comment by o-nikolas** [comment]
Similar to the Dynamic Task Mapping example you brought up, passing data via xcoms between tasks is different depending on if the source/destination is a classic operator or a taskflow operator. It would amazing to have that unified as well.

**Comment by potiuk** [comment]
If anyone wants to attempt to do so - that would be welcome. Unification of user interface is always welcome. We just need to keep backwards compatibility. But I see no reason to reject such a change when someone figures out how to do it if backwards-compatibiliyt is kept.

In short - yes, why not, someone just has to figure out how to do it.
